In [75]:
import pandas as pd

from baseball_scraper import playerid_lookup
from baseball_scraper import statcast_pitcher
from baseball_scraper import batting_stats_range
from baseball_scraper import statcast
from baseball_scraper import statcast_batter

from sklearn.cluster import KMeans, k_means
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import StandardScaler

In [157]:
pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 400)

In [77]:
#data = batting_stats_range('2015-04-01', '2020-10-27')


In [78]:
#data.head()

In [79]:
#data[data['G'] > 100]

In [80]:
# hitters = data[data['G'] > 100]

# hitter_ids = hitters['mlb_ID']

# hitter_ids.to_csv('./veteran_hitter_ids',index=False)

In [81]:
hitter_ids = pd.read_csv('../veteran_hitter_ids')
hitter_ids.drop(columns=['Unnamed: 0'], inplace=True)

In [82]:
hitter_ids

mlb_ID
0    547989
1    554429
2    660670
3    542436
4    642715
..      ...
689  458913
690  605548
691  475582
692  450314
693  572287

[694 rows x 1 columns]

In [83]:
# sc_data = statcast(start_dt='2015-04-01', end_dt= '2020-10-29')

In [84]:
#sc_data.to_csv('./statcast_data_2015.csv', index=False)

In [85]:
sc_data = pd.read_csv('../statcast_data_2015.csv')

/Users/johnnybex/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (45) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [86]:
sc_data['game_date'].isnull().sum()

0

In [87]:
sc_data.shape

(3925778, 91)

In [88]:
sc_data['pitcher'].isnull().sum()

0

In [89]:
#sc_data.dropna(subset=['batter'],inplace=True)

In [90]:
sc_data['batter'] = sc_data['batter'].astype(int)

In [91]:
sc_data

Unnamed: 0  index pitch_type   game_date  release_speed  \
0                 0      0         FF  2020-10-27           96.7   
1                 1      1         FF  2020-10-27           94.1   
2                 2      2         FF  2020-10-27           94.9   
3                 3      3         FF  2020-10-27           94.4   
4                 4      4         CU  2020-10-27           81.4   
...             ...    ...        ...         ...            ...   
3925773     3925773   4036         CU  2015-04-05           75.9   
3925774     3925774   4037         FF  2015-04-05           91.6   
3925775     3925775   4038         FF  2015-04-05           94.8   
3925776     3925776   4039         FF  2015-04-05           93.4   
3925777     3925777   4040         FF  2015-04-05           93.6   

         release_pos_x  release_pos_z  player_name  batter   pitcher  \
0                 1.58           5.99  Julio Urias  642715  628711.0   
1                 2.91           5.45  Julio Urias  642715  628711.0   
2                 1.77           6.02  Julio Urias  642715  628711.0   
3                 1.66           5.93  Julio Urias  670712  628711.0   
4                 1.46           6.06  Julio Urias  670712  628711.0   
...                ...            ...          ...     ...       ...   
3925773           2.18           5.56   Jon Lester  572761  452657.0   
3925774           2.28           5.21   Jon Lester  572761  452657.0   
3925775           2.13           5.31   Jon Lester  572761  452657.0   
3925776           2.12           5.31   Jon Lester  572761  452657.0   
3925777           2.28           5.30   Jon Lester  572761  452657.0   

            events      description  spin_dir  spin_rate_deprecated  \
0        strikeout    called_strike       NaN                   NaN   
1              NaN    called_strike       NaN                   NaN   
2              NaN  swinging_strike       NaN                   NaN   
3        strikeout    called_strike       NaN                   NaN   
4              NaN             ball       NaN                   NaN   
...            ...              ...       ...                   ...   
3925773        NaN             ball       NaN                   NaN   
3925774        NaN             ball       NaN                   NaN   
3925775        NaN             foul       NaN                   NaN   
3925776        NaN  swinging_strike       NaN                   NaN   
3925777        NaN    called_strike       NaN                   NaN   

         break_angle_deprecated  break_length_deprecated  zone  \
0                           NaN                      NaN   4.0   
1                           NaN                      NaN   1.0   
2                           NaN                      NaN   2.0   
3                           NaN                      NaN   4.0   
4                           NaN                      NaN  13.0   
...                         ...                      ...   ...   
3925773                     NaN                      NaN  13.0   
3925774                     NaN                      NaN  13.0   
3925775                     NaN                      NaN   1.0   
3925776                     NaN                      NaN   5.0   
3925777                     NaN                      NaN   6.0   

                                            des game_type stand p_throws  \
0         Willy Adames called out on strikes.           W     R        L   
1                                           NaN         W     R        L   
2                                           NaN         W     R        L   
3        Mike Brosseau called out on strikes.           W     R        L   
4                                           NaN         W     R        L   
...                                         ...       ...   ...      ...   
3925773                                     NaN         R     L        L   
3925774                                     NaN         R     L        L 

In [92]:
# vet_hitter_statcast = sc_data[sc_data['batter'] in hitter_ids.values]
vet_hitter_statcast = pd.merge(sc_data, hitter_ids, left_on= 'batter', right_on='mlb_ID')

In [93]:
vet_hitter_statcast.loc[:, 'game_date'] = pd.to_datetime(vet_hitter_statcast['game_date'],
                                                        format= '%Y-%m-%d')

In [94]:
vet_hitter_statcast[vet_hitter_statcast['game_date'] > pd.to_datetime('2015-11-1')]

Unnamed: 0  index pitch_type  game_date  release_speed  \
0                 0      0         FF 2020-10-27           96.7   
1                 1      1         FF 2020-10-27           94.1   
2                 2      2         FF 2020-10-27           94.9   
3                68     68         SI 2020-10-27          102.1   
4                69     69         SI 2020-10-27          100.5   
...             ...    ...        ...        ...            ...   
3644753     3212271  17511         FF 2016-04-03           94.5   
3644754     3212272  17514         FF 2016-04-03           93.9   
3644755     3212332  17694         SL 2016-04-03           89.4   
3644756     3212333  17698         FF 2016-04-03           95.7   
3644757     3212334  17699         SL 2016-04-03           86.3   

         release_pos_x  release_pos_z       player_name  batter   pitcher  \
0                 1.58           5.99       Julio Urias  642715  628711.0   
1                 2.91           5.45       Julio Urias  642715  628711.0   
2                 1.77           6.02       Julio Urias  642715  628711.0   
3                -1.51           6.51  Brusdar Graterol  642715  660813.0   
4                -1.58           6.69  Brusdar Graterol  642715  660813.0   
...                ...            ...               ...     ...       ...   
3644753          -1.18           6.28      Chris Archer  499624  502042.0   
3644754          -1.24           6.13      Chris Archer  499624  502042.0   
3644755          -1.09           6.24      Chris Archer  499624  502042.0   
3644756          -1.09           6.00      Chris Archer  499624  502042.0   
3644757          -1.37           6.34      Chris Archer  499624  502042.0   

            events      description  spin_dir  spin_rate_deprecated  \
0        strikeout    called_strike       NaN                   NaN   
1              NaN    called_strike       NaN                   NaN   
2              NaN  swinging_strike       NaN                   NaN   
3        field_out    hit_into_play       NaN                   NaN   
4              NaN             ball       NaN                   NaN   
...            ...              ...       ...                   ...   
3644753        NaN             ball       NaN                   NaN   
3644754        NaN             foul       NaN                   NaN   
3644755  strikeout  swinging_strike       NaN                   NaN   
3644756        NaN             foul       NaN                   NaN   
3644757        NaN    called_strike       NaN                   NaN   

         break_angle_deprecated  break_length_deprecated  zone  \
0                           NaN                      NaN   4.0   
1                           NaN                      NaN   1.0   
2                           NaN                      NaN   2.0   
3                           NaN                      NaN   4.0   
4                           NaN                      NaN  11.0   
...                         ...                      ...   ...   
3644753                     NaN                      NaN  11.0   
3644754                     NaN                      NaN   5.0   
3644755                     NaN                      NaN  14.0   
3644756                     NaN                      NaN   9.0   
3644757                     NaN                      NaN   6.0   

                                                       des game_type stand  \
0                    Willy Adames called out on strikes.           W     R   
1                                                      NaN         W     R   
2                                                      NaN         W     R   
3        Willy Adames grounds out to first baseman Max ...         W     R   
4                                                      NaN         W     R   
...                                                    ...       ...   ...   
3644753                                                NaN         R     R   
3644754  

In [95]:
vet_hitter_statcast['game_date'].dtype

dtype('<M8[ns]')

In [96]:
d = {}

for year, sc_year in vet_hitter_statcast.groupby(vet_hitter_statcast.game_date.dt.year):
    d['df_{}'.format(int(year))] = sc_year


In [97]:
d.keys()

dict_keys(['df_2015', 'df_2016', 'df_2017', 'df_2018', 'df_2019', 'df_2020'])

In [98]:
df_2015 = d['df_2015']
df_2016 = d['df_2016']
df_2017 = d['df_2017']
df_2018 = d['df_2018']
df_2019 = d['df_2019']
df_2020 = d['df_2020']

In [99]:
#df = vet_hitter_statcast[vet_hitter_statcast['events'].notna()]
df = pd.concat([df_2015, df_2016, df_2017, df_2018])

In [100]:
df = df[df['events'].notna()]

In [101]:
df.drop('Unnamed: 0', axis=1, inplace=True)

In [102]:
df['pitch_name'].value_counts()

4-Seam Fastball     235353
Slider              112978
2-Seam Fastball      87637
Changeup             77462
Sinker               63135
Curveball            51399
Cutter               38943
Knuckle Curve        16764
Split-Finger         12169
Knuckleball           2815
Intentional Ball      1761
Eephus                 231
Forkball               225
Screwball               30
Pitch Out               29
Unknown                  3
Fastball                 3
Name: pitch_name, dtype: int64

In [103]:
# df['hard'] = 0
# df['soft'] = 0

In [104]:
# def hard_or_soft(pitch):
    
#     hard = ['4-Seam Fastball', '2-Seam Fastball', 'Sinker', 'Cutter', 'Fastball']
#     if pitch 
    


In [105]:
#df.index.to_list()

In [106]:
#df.to_csv('./dummy_statcast_2015-2020.csv')

In [107]:
#df = pd.read_csv('./dummy_statcast_2015-2020.csv')

In [108]:
df.head()

index pitch_type  game_date  release_speed  release_pos_x  \
18572   3084         CH 2015-10-04           88.8          -1.73   
18574  17758         FF 2015-09-29           94.3          -2.11   
18578  18811         FT 2015-09-29           90.8          -1.59   
18585  19904         CH 2015-09-29           82.1          -1.59   
18590   1537         SI 2015-09-27           88.0          -2.29   

       release_pos_z    player_name  batter   pitcher     events  \
18572           5.94  Logan Kensing  571970  434137.0  field_out   
18574           5.93  Cam Bedrosian  571970  592135.0  strikeout   
18578           6.21  Nick Tropeano  571970  607374.0   home_run   
18585           6.35  Nick Tropeano  571970  607374.0  strikeout   
18590           5.51   Chris Heston  571970  518790.0     triple   

               description  spin_dir  spin_rate_deprecated  \
18572        hit_into_play       NaN                   NaN   
18574      swinging_strike       NaN                   NaN   
18578  hit_into_play_score       NaN                   NaN   
18585        called_strike       NaN                   NaN   
18590  hit_into_play_score       NaN                   NaN   

       break_angle_deprecated  break_length_deprecated  zone  \
18572                     NaN                      NaN   4.0   
18574                     NaN                      NaN  11.0   
18578                     NaN                      NaN   4.0   
18585                     NaN                      NaN   6.0   
18590                     NaN                      NaN   4.0   

                                                     des game_type stand  \
18572  Max Muncy flies out to center fielder Brad Mil...         R     L   
18574                  Max Muncy strikes out swinging.           R     L   
18578  Max Muncy homers (3) on a fly ball to center f...         R     L   
18585                 Max Muncy called out on strikes.           R     L   
18590  Max Muncy triples (1) on a fly ball to left fi...         R     L   

      p_throws home_team away_team type  hit_location   bb_type  balls  \
18572        R       SEA       OAK    X           8.0  fly_ball    0.0   
18574        R       LAA       OAK    S           2.0       NaN    1.0   
18578        R       LAA       OAK    X           NaN  fly_ball    3.0   
18585        R       LAA       OAK    S           2.0       NaN    2.0   
18590        R       OAK        SF    X           7.0  fly_ball    3.0   

       strikes  game_year  pfx_x  pfx_z  plate_x  plate_z  on_3b     on_2b  \
18572      1.0     2015.0  -1.37   0.31    -0.39     2.55    NaN       NaN   
18574      2.0     2015.0  -0.53   1.83    -0.85     3.03    NaN  456714.0   
18578      2.0     2015.0  -1.17   1.48    -0.68     2.49    NaN       NaN   
18585      2.0     2015.0  -0.86   0.55     0.69     2.73    NaN       NaN   
18590      2.0     2015.0  -1.29   0.55    -0.45     2.26    NaN       NaN   

          on_1b  outs_when_up  inning inning_topbot    hc_x   hc_y  \
18572       NaN           2.0     8.0           Top  152.97  48.79   
18574  543434.0           2.0     7.0           Top     NaN    NaN   
18578       NaN           2.0     5.0           Top  114.48  30.29   
18585       NaN           0.0     3.0           Top     NaN    NaN   
18590  502210.0           0.0     6.0           Bot   41.12  84.94   

       tfs_deprecated  tfs_zulu_deprecated  fielder_2  umpire          sv_id  \
18572             NaN                  NaN   491696.0     NaN  151004_143649   
18574             NaN                  NaN   542208.0     NaN  150929_212741   
18578             NaN                  NaN   542208.0     NaN  150929_203813   
18585             NaN                  NaN   542208.0     NaN  150929_194649   
18590             NaN                  NaN   623143.0     NaN  150927_145447   

         vx0      vy0    vz0      ax      ay      az  sz_top  sz_bot  \
18572  6.299 -129.186 -2.466 -14.744  32.371 -30.811    3.37    1.53   
18574  4.448 -

In [109]:
print(df['spin_dir'].isna().sum())
print(df['spin_rate_deprecated'].isna().sum())
print(df['break_angle_deprecated'].isna().sum())
print(df['break_length_deprecated'].isna().sum())
print(df['tfs_deprecated'].isna().sum())
print(df['tfs_zulu_deprecated'].isna().sum())
print(df['umpire'].isna().sum())
print(df['sv_id'].isna().sum())

703671
703671
703671
703671
703671
703671
703671
2277


In [110]:
print(len(df))

703671


In [111]:
df.drop(columns=['spin_dir', 'spin_rate_deprecated', 'break_angle_deprecated',
                'break_length_deprecated', 'tfs_deprecated', 'tfs_zulu_deprecated',
                'umpire'], inplace=True)

In [112]:
df.head()

index pitch_type  game_date  release_speed  release_pos_x  \
18572   3084         CH 2015-10-04           88.8          -1.73   
18574  17758         FF 2015-09-29           94.3          -2.11   
18578  18811         FT 2015-09-29           90.8          -1.59   
18585  19904         CH 2015-09-29           82.1          -1.59   
18590   1537         SI 2015-09-27           88.0          -2.29   

       release_pos_z    player_name  batter   pitcher     events  \
18572           5.94  Logan Kensing  571970  434137.0  field_out   
18574           5.93  Cam Bedrosian  571970  592135.0  strikeout   
18578           6.21  Nick Tropeano  571970  607374.0   home_run   
18585           6.35  Nick Tropeano  571970  607374.0  strikeout   
18590           5.51   Chris Heston  571970  518790.0     triple   

               description  zone  \
18572        hit_into_play   4.0   
18574      swinging_strike  11.0   
18578  hit_into_play_score   4.0   
18585        called_strike   6.0   
18590  hit_into_play_score   4.0   

                                                     des game_type stand  \
18572  Max Muncy flies out to center fielder Brad Mil...         R     L   
18574                  Max Muncy strikes out swinging.           R     L   
18578  Max Muncy homers (3) on a fly ball to center f...         R     L   
18585                 Max Muncy called out on strikes.           R     L   
18590  Max Muncy triples (1) on a fly ball to left fi...         R     L   

      p_throws home_team away_team type  hit_location   bb_type  balls  \
18572        R       SEA       OAK    X           8.0  fly_ball    0.0   
18574        R       LAA       OAK    S           2.0       NaN    1.0   
18578        R       LAA       OAK    X           NaN  fly_ball    3.0   
18585        R       LAA       OAK    S           2.0       NaN    2.0   
18590        R       OAK        SF    X           7.0  fly_ball    3.0   

       strikes  game_year  pfx_x  pfx_z  plate_x  plate_z  on_3b     on_2b  \
18572      1.0     2015.0  -1.37   0.31    -0.39     2.55    NaN       NaN   
18574      2.0     2015.0  -0.53   1.83    -0.85     3.03    NaN  456714.0   
18578      2.0     2015.0  -1.17   1.48    -0.68     2.49    NaN       NaN   
18585      2.0     2015.0  -0.86   0.55     0.69     2.73    NaN       NaN   
18590      2.0     2015.0  -1.29   0.55    -0.45     2.26    NaN       NaN   

          on_1b  outs_when_up  inning inning_topbot    hc_x   hc_y  fielder_2  \
18572       NaN           2.0     8.0           Top  152.97  48.79   491696.0   
18574  543434.0           2.0     7.0           Top     NaN    NaN   542208.0   
18578       NaN           2.0     5.0           Top  114.48  30.29   542208.0   
18585       NaN           0.0     3.0           Top     NaN    NaN   542208.0   
18590  502210.0           0.0     6.0           Bot   41.12  84.94   623143.0   

               sv_id    vx0      vy0    vz0      ax      ay      az  sz_top  \
18572  151004_143649  6.299 -129.186 -2.466 -14.744  32.371 -30.811    3.37   
18574  150929_212741  4.448 -137.085 -5.965  -5.444  28.396 -10.329    3.37   
18578  150929_203813  4.944 -132.094 -6.639 -13.380  25.744 -16.191    3.37   
18585  150929_194649  6.909 -119.291 -2.345  -7.418  25.842 -28.675    3.37   
18590  150927_145447  7.450 -127.963 -2.810 -14.045  21.256 -27.920    3.36   

       sz_bot  hit_distance_sc  launch_speed  launch_angle  effective_speed  \
18572    1.53            364.0          94.1          30.0             87.9   
18574    1.53              NaN           NaN           NaN             94.7   
18578    1.53            413.0         100.8          29.0             89.1   
18585    1.53              NaN           NaN           NaN             79.5   
18590    1.54            359.0          99.2          32.0             88.6   

       release_spin_rate  release_extension   game_pk  pitcher.1  fielder_2.1  \
18572             1757.0                6.2  416075.0   434137.0     491696.0   
18

In [113]:
df['babip_value'].value_counts()

0.0    563398
1.0    140273
Name: babip_value, dtype: int64

In [114]:
df['iso_value'].value_counts()

0.0    646195
1.0     32218
3.0     21925
2.0      3333
Name: iso_value, dtype: int64

In [115]:
df['woba_value'].value_counts()

0.00    473175
0.90    110802
0.70     62218
1.25     32218
2.00     21925
1.60      3333
Name: woba_value, dtype: int64

wOBA = (0.690×uBB + 0.722×HBP + 0.888×1B + 1.271×2B + 1.616×3B +
2.101×HR) / (AB + BB – IBB + SF + HBP)

In [116]:
interesting_columns = ['batter', 'zone', 'game_type', 'stand', 'p_throws', 'hit_location', 'events'
                  'bb_type', 'balls', 'strikes', 'on_3b', 'on_2b', 'on_1b',
                  'outs_when_up', 'inning', 'inning_topbot', 'hc_x', 'hc_y',
                  'hit_distance_sc', 'launch_speed', 'launch_angle', 'effective_speed',
                  'release_spin_rate', 'estimated_ba_using_speedangle', 
                   'estimated_woba_using_speedangle', 'woba_value','babip_value', 
                   'iso_value', 'launch_speed_angle', 'pitch_number', 'hard', 'soft']

In [117]:
hitting_cols = ['batter', 'zone', 'game_type', 'stand', 'p_throws', 'hit_location', 
                'events','bb_type', 'balls', 'strikes', 'on_3b', 'on_2b', 'on_1b',
                'outs_when_up', 'inning', 'hc_x', 'hc_y','hit_distance_sc', 
                'launch_speed', 'launch_angle', 'estimated_ba_using_speedangle', 
                'woba_value','babip_value', 'iso_value', 'launch_speed_angle', 
                'pitch_number', 'pitch_name', 'effective_speed','if_fielding_alignment', 
                'of_fielding_alignment', 'bat_score', 'fld_score']

In [118]:
sc_hitting = ['batter', 'game_type', 'p_throws', 'bb_type', 
            'hc_x', 'hc_y','hit_distance_sc', 'launch_speed', 'launch_angle', 
            'woba_value','babip_value', 'iso_value', 'launch_speed_angle', 'pitch_name']

**Curious about ISO value**

In [119]:
hitting_df = df[sc_hitting]

In [120]:
#hitting_df = pd.get_dummies(data=hitting_df, columns=['events'])

In [121]:
hard = ['4-Seam Fastball', '2-Seam Fastball', 'Sinker', 'Cutter', 'Fastball']

# for index in df.index.to_list():
#     #print(index)
#     if df.loc[index, 'pitch_name'] in hard:
#         df.loc[index, 'hard'] = 1
#     else:
#          df.loc[index, 'soft'] = 1
            
hitting_df['hard'] = [1 if x in hard else 0 for x in hitting_df['pitch_name']]
#hitting_df['soft'] = [1 if x not in hard else 0 for x in hitting_df['pitch_name']]
hitting_df.drop(columns=['pitch_name'], inplace=True)

<ipython-input-121-e0deca61c74b>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hitting_df['hard'] = [1 if x in hard else 0 for x in hitting_df['pitch_name']]
/Users/johnnybex/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


Fill in players on base with 1 instead of player_id, null is 0


In [122]:
hitting_df = pd.get_dummies(data=hitting_df, columns=['bb_type'], drop_first=True)

In [123]:
hitting_df = pd.get_dummies(data=hitting_df, columns=['launch_speed_angle'], drop_first=True)

In [124]:
# hitting_df['winning'] = hitting_df['bat_score'] - hitting_df['fld_score']
# hitting_df['winning'] = ['losing' if x <0 else 'tied' if x==0 else 'winning' for x in hitting_df['winning']]
# hitting_df = pd.get_dummies(data=hitting_df, columns=['winning'], drop_first=True)
# hitting_df = hitting_df.drop(columns=['bat_score', 'fld_score'])

In [125]:
# Separating Regualar season

hitting_df['regular_season'] = [1 if x == 'R' else 0 for x in df['game_type']]

In [126]:
# hitting_df = pd.get_dummies(data = hitting_df, columns = ['if_fielding_alignment', 'of_fielding_alignment'], 
#                     drop_first=True)

In [127]:
# Separating postseason

postseason = ['F', 'D', 'L', 'W']

hitting_df['postseason'] = [1 if x in postseason else 0 for x in df['game_type']]

In [128]:
# Binarizing hitting handedness

#hitting_df = pd.get_dummies(data=hitting_df, columns=['stand'], drop_first=True)

#df['stance'] = [1 if x=='R' else 0 for x in df['stand']]

In [129]:
hitting_df = pd.get_dummies(data=hitting_df, columns=['p_throws'], drop_first=True)

In [130]:
#df['half_inning'] = [0 if x == 'Top' else 1 for x in df['inning_topbot']]

In [131]:
# hitting_df[['on_1b', 'on_2b', 'on_3b']].fillna(0)

# hitting_df['on_1b'] = [1 if x > 0 else 0 for x in df['on_1b']]
# hitting_df['on_2b'] = [1 if x > 0 else 0 for x in df['on_2b']]
# hitting_df['on_3b'] = [1 if x > 0 else 0 for x in df['on_3b']]

In [132]:
regular_hitting = hitting_df[hitting_df['game_type'] == 'R']
regular_hitting = regular_hitting.drop(columns=['postseason', 'regular_season', 'game_type'])

In [133]:
regular_hitting.to_csv('./regular_season_sc_hitting')

In [134]:
regular_hitting

batter    hc_x    hc_y  hit_distance_sc  launch_speed  launch_angle  \
18572    571970  152.97   48.79            364.0          94.1          30.0   
18574    571970     NaN     NaN              NaN           NaN           NaN   
18578    571970  114.48   30.29            413.0         100.8          29.0   
18585    571970     NaN     NaN              NaN           NaN           NaN   
18590    571970   41.12   84.94            359.0          99.2          32.0   
...         ...     ...     ...              ...           ...           ...   
3359494  471083  182.28  110.40            266.0          82.5          23.0   
3359495  471083   92.07   71.51            310.0          86.4          31.0   
3359497  471083  107.93  154.86             73.0          75.3           6.0   
3359498  471083  129.13  183.33             11.0          94.0         -13.0   
3359501  471083  141.24  154.86             71.0          72.9           6.0   

         woba_value  babip_value  iso_value  hard  bb_type_ground_ball  \
18572           0.0          0.0        0.0     0                    0   
18574           0.0          0.0        0.0     1                    0   
18578           2.0          0.0        3.0     1                    0   
18585           0.0          0.0        0.0     0                    0   
18590           1.6          1.0        2.0     1                    0   
...             ...          ...        ...   ...                  ...   
3359494         0.0          0.0        0.0     1                    0   
3359495         0.0          0.0        0.0     1                    0   
3359497         0.0          0.0        0.0     1                    1   
3359498         0.0          0.0        0.0     1                    1   
3359501         0.0          0.0        0.0     1                    1   

         bb_type_line_drive  bb_type_popup  launch_speed_angle_2.0  \
18572                     0              0                       0   
18574                     0              0                       0   
18578                     0              0                       0   
18585                     0              0                       0   
18590                     0              0                       0   
...                     ...            ...                     ...   
3359494                   1              0                       0   
3359495                   0              0                       0   
3359497                   0              0                       1   
3359498                   0              0                       1   
3359501                   0              0                       1   

         launch_speed_angle_3.0  launch_speed_angle_4.0  \
18572                         1                       0   
18574                         0                       0   
18578                         0                       0   
18585                         0                       0   
18590                         0                       0   
...                         ...                     ...   
3359494                       1                       0   
3359495                       1                       0   
3359497                       0                       0   
3359498                       0                       0   
3359501                       0                       0   

         launch_speed_angle_5.0  launch_speed_angle_6.0  p_throws_R  
18572                         0                       0           1  
18574                         0                       0           1  
18578                         0                       1           1  
18585                         0                       0           1  
18590                         1                       0           1  
...                         ...                     ...         ...  
3359494                       0                       0           1  
3359495                       0                  

In [135]:
rh_gb = regular_hitting.groupby('batter').mean()

In [136]:
rh_gb

hc_x        hc_y  hit_distance_sc  launch_speed  launch_angle  \
batter                                                                        
116338  125.443848  127.028242       162.284182     88.406295     10.786925   
120074  140.790445  112.335695       205.639136     93.089468     16.281993   
121347  121.053223  120.489121       174.983903     90.854797     11.721402   
133380  123.462282  119.972500       184.731183     87.362562     15.761084   
134181  124.160108  119.199850       181.657124     89.018396     13.367310   
...            ...         ...              ...           ...           ...   
664057  128.141053  131.789053       139.686047     87.140230      5.034483   
664058  120.845906  127.307756       161.290598     84.000000     13.404000   
664068  123.722115  121.515740       185.387755     86.171118     17.394410   
665742  126.675270  124.422317       148.747331     90.484466      6.288026   
666971  123.933264  119.926943       177.146893     91.076842     11.126316   

        woba_value  babip_value  iso_value      hard  bb_type_ground_ball  \
batter                                                                      
116338    0.309315     0.181019   0.154657  0.620387             0.365554   
120074    0.396011     0.191781   0.251410  0.622885             0.258662   
121347    0.332737     0.154734   0.181293  0.650115             0.279446   
133380    0.308607     0.193424   0.162476  0.647969             0.299807   
134181    0.365899     0.225458   0.172851  0.637388             0.325035   
...            ...          ...        ...       ...                  ...   
664057    0.252632     0.177632   0.046053  0.565789             0.348684   
664058    0.303247     0.250000   0.081169  0.597403             0.340909   
664068    0.267593     0.193416   0.104938  0.553498             0.246914   
665742    0.395361     0.204124   0.191753  0.579381             0.356701   
666971    0.330916     0.225191   0.156489  0.496183             0.328244   

        bb_type_line_drive  bb_type_popup  launch_speed_angle_2.0  \
batter                                                              
116338            0.154657       0.052724                0.288225   
120074            0.209508       0.061241                0.186946   
121347            0.153580       0.043880                0.220554   
133380            0.232108       0.069632                0.234043   
134181            0.222170       0.058713                0.248473   
...                    ...            ...                     ...   
664057            0.164474       0.013158                0.230263   
664058            0.253247       0.074675                0.275974   
664068            0.181070       0.053498                0.185185   
665742            0.138144       0.032990                0.253608   
666971            0.221374       0.022901                0.232824   

        launch_speed_angle_3.0  launch_speed_angle_4.0  \
batter                                                   
116338                0.181019                0.159930   
120074                0.186140                0.165995   
121347                0.154734                0.130485   
133380                0.241779                0.185687   
134181                0.201503                0.205261   
...                        ...                     ...   
664057                0.078947                0.184211   
664058                0.233766                0.256494   
664068                0.205761                0.174897   
665742                0.103093                0.152577   
666971                0.171756                0.194656   

        launch_speed_angle_5.0  launch_speed_angle_6.0  p_throws_R  
batter                                                              
116338                0.038664                0.036907    0.695958  
120074                0.059629                0.103948    0.728445  
121347                0.046189                0.062356 

In [137]:
# rh_gb = regular_hitting.groupby(by=['batter', 'hard', 'zone','outs_when_up', 'inning', 
#                             'balls', 'strikes'], as_index=False).mean()

In [138]:
# rh_gb.fillna(0, inplace=True)

In [139]:
rh_gb_sc = StandardScaler().fit_transform(rh_gb)

In [149]:
model = KMeans(n_clusters=3, random_state=0)
model.fit(rh_gb_sc)

KMeans(n_clusters=3, random_state=0)

In [150]:
labels = model.labels_

rh_gb['labels'] = labels
rh_gb.head()

hc_x        hc_y  hit_distance_sc  launch_speed  launch_angle  \
batter                                                                        
116338  125.443848  127.028242       162.284182     88.406295     10.786925   
120074  140.790445  112.335695       205.639136     93.089468     16.281993   
121347  121.053223  120.489121       174.983903     90.854797     11.721402   
133380  123.462282  119.972500       184.731183     87.362562     15.761084   
134181  124.160108  119.199850       181.657124     89.018396     13.367310   

        woba_value  babip_value  iso_value      hard  bb_type_ground_ball  \
batter                                                                      
116338    0.309315     0.181019   0.154657  0.620387             0.365554   
120074    0.396011     0.191781   0.251410  0.622885             0.258662   
121347    0.332737     0.154734   0.181293  0.650115             0.279446   
133380    0.308607     0.193424   0.162476  0.647969             0.299807   
134181    0.365899     0.225458   0.172851  0.637388             0.325035   

        bb_type_line_drive  bb_type_popup  launch_speed_angle_2.0  \
batter                                                              
116338            0.154657       0.052724                0.288225   
120074            0.209508       0.061241                0.186946   
121347            0.153580       0.043880                0.220554   
133380            0.232108       0.069632                0.234043   
134181            0.222170       0.058713                0.248473   

        launch_speed_angle_3.0  launch_speed_angle_4.0  \
batter                                                   
116338                0.181019                0.159930   
120074                0.186140                0.165995   
121347                0.154734                0.130485   
133380                0.241779                0.185687   
134181                0.201503                0.205261   

        launch_speed_angle_5.0  launch_speed_angle_6.0  p_throws_R  labels  
batter                                                                      
116338                0.038664                0.036907    0.695958       2  
120074                0.059629                0.103948    0.728445       0  
121347                0.046189                0.062356    0.680139       0  
133380                0.052224                0.044487    0.762089       0  
134181                0.052607                0.051667    0.721935       0

In [151]:
score = silhouette_score(rh_gb, labels, metric='euclidean')
score

0.29502481834206346

In [152]:
data = batting_stats_range('2015-04-01', '2018-11-07')

data = data[data['G'] > 100]

data.head()

Name  Age  #days     Lev                Tm    G    PA    AB  \
3        José Abreu   31    893  MLB-AL           Chicago  597  2591  2357   
4     Dustin Ackley   28   1733  MLB-AL  New York,Seattle  125   334   299   
5  Ronald Acuna Jr.   20    879  MLB-NL           Atlanta  111   487   433   
6  Cristhian Adames   25   1399  MLB-NL          Colorado  152   328   291   
9        Lane Adams   28    879  MLB-NL           Atlanta  108   151   134   

     R    H   2B  3B   HR  RBI   BB  IBB   SO  HBP  SH  SF  GDP  SB  CS  \
3  318  682  145  11  110  381  158   31  493   56   0  20   72   5   0   
4   34   64   11   3   10   34   26    0   54    1   3   5    3   2   2   
5   78  127   26   4   26   64   45    2  123    6   0   3    4  16   4   
6   30   62    8   4    2   20   28    1   64    5   4   0    5   2   3   
9   29   36    5   1    7   26   14    0   45    1   1   1    4  11   0   

      BA    OBP    SLG    OPS  mlb_ID  
3  0.289  0.346  0.500  0.846  547989  
4  0.214  0.275  0.371  0.646  554429  
5  0.293  0.366  0.552  0.917  660670  
6  0.213  0.293  0.289  0.582  542436  
9  0.269  0.340  0.478  0.818  572669

In [153]:
group_0 = rh_gb[rh_gb['labels'] == 0]

group_1 = rh_gb[rh_gb['labels'] == 1]

group_2 = rh_gb[rh_gb['labels'] == 2]

group_3 = rh_gb[rh_gb['labels'] == 3]

In [158]:
pd.merge(group_0, data, left_on='batter', right_on='mlb_ID')

hc_x        hc_y  hit_distance_sc  launch_speed  launch_angle  \
0    140.790445  112.335695       205.639136     93.089468     16.281993   
1    121.053223  120.489121       174.983903     90.854797     11.721402   
2    123.462282  119.972500       184.731183     87.362562     15.761084   
3    124.160108  119.199850       181.657124     89.018396     13.367310   
4    130.831298  119.803543       182.181650     89.447312     13.723739   
5    123.016267  119.978836       187.299728     90.227561     14.841463   
6    131.132398  119.434975       182.100749     88.594583     14.466376   
7    118.365036  121.023116       180.668585     90.314640     13.296794   
8    128.653632  121.699372       179.070000     86.447307     13.133489   
9    129.454293  124.155152       164.855093     91.309936      9.765605   
10   132.552148  122.808960       177.065421     88.995719     14.179795   
11   132.008548  115.804762       185.580484     93.072791     11.911628   
12   131.064444  119.064556       178.297891     88.399757     12.219611   
13   124.903378  123.301757       198.564767     87.564115     17.110048   
14   135.009765  120.888054       178.486792     89.517586     13.075862   
15   121.313738  124.043095       175.882192     87.241748     13.963592   
16   130.640036  122.024508       165.533973     89.811547      7.557365   
17   125.396112  125.056799       180.656184     86.654259     15.001852   
18   127.949427  125.077260       161.116173     88.420842      9.962025   
19   131.568359  125.709297       181.539589     85.744236     16.884718   
20   116.130859  120.405298       187.422827     90.007329     16.696217   
21   139.656757  113.518696       202.872340     91.601468     15.611009   
22   121.896883  119.586265       177.743056     90.984277      9.477987   
23   122.702479  119.506880       186.175532     89.474888     11.434978   
24   126.291392  123.225095       172.131387     89.448232     11.241158   
25   115.295774  121.496847       188.791290     90.618116     16.779295   
26   123.095302  123.045637       172.883639     88.001309     12.915663   
27   121.617349  120.187070       189.610320     87.330346     14.977987   
28   126.106891  126.825507       156.745322     89.298315      9.148876   
29   140.609464  117.294788       199.505435     88.437214     18.554286   
30   122.634651  118.764651       191.390625     89.632367     14.246377   
31   124.310744  123.049194       162.945471     89.994427      9.873626   
32   130.688996  116.943936       181.011258     92.046004     10.268456   
33   116.796641  120.698804       191.337037     88.904486     18.527352   
34   120.566330  120.458296       190.480660     86.211175     18.358628   
35   141.548768  120.539575       184.970732     88.420800     17.752889   
36   132.067543  117.774423       185.529506     88.081124     15.640449   
37   117.978347  123.539927       174.771946     87.426821     16.052980   
38   122.472838  116.919816       184.456782     93.580829     11.404447   
39   123.838306  119.571064       178.344641     91.697871     13.072115   
40   121.740662  120.396118       183.468627     86.267384     16.286738   
41   136.926986  120.126849       183.654206     86.181742     11.168539   
42   135.139260  112.792446       211.449604     90.539337     19.031056   
43   130.465840  119.656396       179.080292     86.936164     14.624046   
44   122.145326  120.252505       183.759292     88.913534     15.024168   
45   114.759374  121.897677       177.864919     85.933333     15.750693   
46   121.336798  127.157217       171.388970     86.445288     14.662304   
47   121.986779  124.718539       182.865772     88.334585     12.893281   
48   121.387512  118.929836       175.411702     87.667248     13.148256   
49   136.934763  111.275521       203.517182     90.763133     16.311093   
50   133.387991  122.231224       179.334465     87.340062     12.691834   
51   123.844067  121.038935       170.021136     90.748789 

In [155]:
pd.merge(group_1, data, left_on='batter', right_on='mlb_ID')

hc_x        hc_y  hit_distance_sc  launch_speed  launch_angle  \
0   126.783293  145.826886       115.450000     88.776119      0.925373   
1   130.381147  146.760596        95.564706     83.356425     -2.754190   
2   138.473571  139.117269       120.064356     85.281333      3.128889   
3   126.035448  150.851269       129.703297     78.873000     12.420000   
4   127.365088  141.677632       145.566265     85.420408     10.051020   
5   133.597757  158.843178        79.971429     84.525316     -0.177215   
6   129.826030  149.659950       127.847826     83.704138      3.958621   
7   129.838788  134.229697       121.737864     84.518807      3.821101   
8   132.810379  154.905909       106.096386     83.152941      3.725490   
9   129.122685  145.127584       107.560976     82.574803      2.614173   
10  125.555194  148.532171       100.022989     82.397170     -1.141509   
11  129.578940  132.797881       125.481752     89.615753      2.712329   
12  125.058025  150.844074        87.630252     84.150725     -4.362319   
13  122.481833  141.047222       121.117241     84.324359      5.942308   

    woba_value  babip_value  iso_value      hard  bb_type_ground_ball  \
0     0.136594     0.083333   0.043478  0.753623             0.442029   
1     0.203994     0.172524   0.019169  0.680511             0.511182   
2     0.273620     0.168712   0.085890  0.702454             0.453988   
3     0.115918     0.085714   0.016327  0.632653             0.322449   
4     0.196784     0.152047   0.029240  0.631579             0.374269   
5     0.100235     0.093897   0.004695  0.591549             0.384977   
6     0.145686     0.113725   0.015686  0.709804             0.533333   
7     0.273171     0.182927   0.067073  0.649390             0.442073   
8     0.094177     0.080321   0.012048  0.718876             0.361446   
9     0.168103     0.125000   0.034483  0.655172             0.456897   
10    0.203196     0.168950   0.036530  0.680365             0.438356   
11    0.215650     0.154472   0.040650  0.560976             0.373984   
12    0.197854     0.171674   0.047210  0.695279             0.527897   
13    0.192509     0.157303   0.026217  0.666667             0.408240   

    bb_type_line_drive  bb_type_popup  launch_speed_angle_2.0  \
0             0.090580       0.014493                0.278986   
1             0.127796       0.015974                0.316294   
2             0.153374       0.046012                0.365031   
3             0.102041       0.053061                0.142857   
4             0.140351       0.052632                0.239766   
5             0.065728       0.037559                0.234742   
6             0.125490       0.027451                0.298039   
7             0.134146       0.024390                0.335366   
8             0.076305       0.052209                0.212851   
9             0.068966       0.021552                0.301724   
10            0.105023       0.018265                0.269406   
11            0.117886       0.028455                0.300813   
12            0.090129       0.042918                0.381974   
13            0.142322       0.041199                0.284644   

    launch_speed_angle_3.0  launch_speed_angle_4.0  launch_speed_angle_5.0  \
0                 0.068841                0.090580                0.014493   
1                 0.063898                0.146965                0.006390   
2                 0.122699                0.156442                0.012270   
3                 0.126531                0.110204                0.000000   
4                 0.140351                0.169591                0.005848   
5                 0.061033                0.065728                0.000000   
6                 0.109804                0.149020                0.007843   
7                 0.109756                0.155488                0.006098   
8                 0.092369                0.084337                0.004016   
9                 0.120690      

In [156]:
pd.merge(group_2, data, left_on='batter', right_on='mlb_ID')

hc_x        hc_y  hit_distance_sc  launch_speed  launch_angle  \
0    125.443848  127.028242       162.284182     88.406295     10.786925   
1    126.693322  128.348288       159.035573     87.498571      7.785714   
2    133.160286  128.649529       156.731203     86.213652      9.293515   
3    136.948891  127.859909       168.240166     85.568821     15.300380   
4    125.322622  125.574355       166.527981     88.473077     10.373932   
5    126.182249  136.548272       128.737120     82.912107      5.226392   
6    135.490972  122.285995       176.574018     87.704093     12.752558   
7    118.441681  123.882090       149.976587     90.236753      4.698710   
8    130.580777  122.165777       162.494340     89.624221      9.605536   
9    128.755545  128.417890       153.101545     85.727309      9.635207   
10   120.930911  127.864393       167.177677     84.599151     14.072187   
11   130.583025  128.947683       164.396244     84.678554     14.870994   
12   130.488695  126.664947       172.689252     85.102620     13.604803   
13   124.197097  123.870985       173.517595     87.831233     13.157534   
14   122.057407  121.117191       178.781570     88.304127     15.142857   
15   129.810643  125.977853       162.659284     86.441260     12.513386   
16   121.716131  125.156490       169.035545     87.936989     11.535484   
17   126.912844  137.803365       150.030120     87.887097     10.650538   
18   124.418426  122.491563       172.496512     87.536717     12.366009   
19   130.391543  123.698620       168.808011     90.202926     10.301527   
20   131.194101  121.578791       163.429245     90.837704      8.181175   
21   120.846886  126.291205       164.293651     88.325463     10.649306   
22   132.392521  126.332773       161.134884     86.827234     12.680851   
23   135.063970  125.569724       165.635036     87.310129     11.909483   
24   131.463035  133.795495       135.230189     85.485859      7.599327   
25   129.989358  134.437055       140.670330     84.710808      6.747502   
26   135.461094  131.385434       141.050209     88.861069      5.545802   
27   130.041045  130.269750       156.354839     86.962998      8.716628   
28   128.531540  126.901600       170.176471     86.407056     11.209677   
29   131.414121  130.666319       144.834395     90.319663      2.078652   
30   125.401400  126.430058       168.580297     85.976792     11.956258   
31   126.155268  128.046436       162.500561     86.019620     12.670671   
32   134.652857  128.045539       137.204280     89.743617      2.037234   
33   115.033324  131.095389       153.559524     83.372423     12.133705   
34   125.574742  131.846392       146.141026     85.578261      5.260870   
35   130.955890  129.225003       138.627193     90.043487      3.126974   
36   137.041422  123.005178       172.407666     87.180599     11.195584   
37   129.433196  122.091910       159.044314     90.181352      6.043091   
38   126.613915  129.684980       157.814815     84.264337     10.853571   
39   124.389009  130.734165       136.467153     85.577586      6.096635   
40   128.753227  126.518633       152.904969     88.137037      8.710950   
41   125.380167  131.037741       150.292017     84.181609     11.256705   
42   127.541046  125.148154       159.741379     86.968153      9.083067   
43   126.841832  133.345516       138.970874     85.822517      9.693157   
44   128.409368  128.732471       157.853195     88.815198      9.815830   
45   134.957130  126.006163       158.565010     88.722382      9.079177   
46   128.087990  126.020766       163.127168     88.348020      9.000000   
47   129.161907  122.360581       164.872227     87.043854     11.779070   
48   130.152456  126.168162       144.916031     89.528882      4.522255   
49   132.078887  127.278506       158.284848     86.323761      9.666667   
50   127.227432  131.435925       139.716535     84.220141      7.197880   
51   131.057810  127.323861       157.258193     86.479947 

In [148]:
pd.merge(group_3, data, left_on='batter', right_on='mlb_ID')

hc_x        hc_y  hit_distance_sc  launch_speed  launch_angle  \
0   126.783293  145.826886       115.450000     88.776119      0.925373   
1   130.381147  146.760596        95.564706     83.356425     -2.754190   
2   138.473571  139.117269       120.064356     85.281333      3.128889   
3   126.035448  150.851269       129.703297     78.873000     12.420000   
4   127.365088  141.677632       145.566265     85.420408     10.051020   
5   133.597757  158.843178        79.971429     84.525316     -0.177215   
6   129.826030  149.659950       127.847826     83.704138      3.958621   
7   129.838788  134.229697       121.737864     84.518807      3.821101   
8   132.810379  154.905909       106.096386     83.152941      3.725490   
9   129.122685  145.127584       107.560976     82.574803      2.614173   
10  125.555194  148.532171       100.022989     82.397170     -1.141509   
11  129.578940  132.797881       125.481752     89.615753      2.712329   
12  125.058025  150.844074        87.630252     84.150725     -4.362319   
13  122.481833  141.047222       121.117241     84.324359      5.942308   

    woba_value  babip_value  iso_value      hard  bb_type_ground_ball  \
0     0.136594     0.083333   0.043478  0.753623             0.442029   
1     0.203994     0.172524   0.019169  0.680511             0.511182   
2     0.273620     0.168712   0.085890  0.702454             0.453988   
3     0.115918     0.085714   0.016327  0.632653             0.322449   
4     0.196784     0.152047   0.029240  0.631579             0.374269   
5     0.100235     0.093897   0.004695  0.591549             0.384977   
6     0.145686     0.113725   0.015686  0.709804             0.533333   
7     0.273171     0.182927   0.067073  0.649390             0.442073   
8     0.094177     0.080321   0.012048  0.718876             0.361446   
9     0.168103     0.125000   0.034483  0.655172             0.456897   
10    0.203196     0.168950   0.036530  0.680365             0.438356   
11    0.215650     0.154472   0.040650  0.560976             0.373984   
12    0.197854     0.171674   0.047210  0.695279             0.527897   
13    0.192509     0.157303   0.026217  0.666667             0.408240   

    bb_type_line_drive  bb_type_popup  launch_speed_angle_2.0  \
0             0.090580       0.014493                0.278986   
1             0.127796       0.015974                0.316294   
2             0.153374       0.046012                0.365031   
3             0.102041       0.053061                0.142857   
4             0.140351       0.052632                0.239766   
5             0.065728       0.037559                0.234742   
6             0.125490       0.027451                0.298039   
7             0.134146       0.024390                0.335366   
8             0.076305       0.052209                0.212851   
9             0.068966       0.021552                0.301724   
10            0.105023       0.018265                0.269406   
11            0.117886       0.028455                0.300813   
12            0.090129       0.042918                0.381974   
13            0.142322       0.041199                0.284644   

    launch_speed_angle_3.0  launch_speed_angle_4.0  launch_speed_angle_5.0  \
0                 0.068841                0.090580                0.014493   
1                 0.063898                0.146965                0.006390   
2                 0.122699                0.156442                0.012270   
3                 0.126531                0.110204                0.000000   
4                 0.140351                0.169591                0.005848   
5                 0.061033                0.065728                0.000000   
6                 0.109804                0.149020                0.007843   
7                 0.109756                0.155488                0.006098   
8                 0.092369                0.084337                0.004016   
9                 0.120690      